extract_feature_region.py

```diff
@@ -0,0 +1,156 @@
+from detectron2.config import get_cfg
+from detectron2.engine import default_setup
+from detectron2.evaluation import inference_context
+from detectron2.modeling import build_model
+from detectron2.modeling import postprocessing
+
+from grid_feats import (
+    add_attribute_config,
+    build_detection_test_loader_with_attributes,
+    build_detection_test_loader_for_images,
+)
+
+# A simple mapper from object detection dataset to VQA dataset names
+dataset_to_folder_mapper = {}
+dataset_to_folder_mapper["coco_2014_train"] = "train2014"
+dataset_to_folder_mapper["coco_2014_val"] = "val2014"
+dataset_to_folder_mapper["visual_genome"] = "visual_genome"
+# One may need to change the Detectron2 code to support coco_2015_test
+# insert "coco_2015_test": ("coco/test2015", "coco/annotations/image_info_test2015.json"),
+# at: https://github.com/facebookresearch/detectron2/blob/master/detectron2/data/datasets/builtin.py#L36
+dataset_to_folder_mapper["coco_2015_test"] = "test2015"
+
+
+def extract_feature_argument_parser():
+    parser = argparse.ArgumentParser(description="Region feature extraction")
+    parser.add_argument(
+        "--config-file", default="", metavar="FILE", help="path to config file"
+    )
+    parser.add_argument(
+        "--dataset",
+        help="name of the dataset",
+        default="coco_2014_train",
+        choices=["coco_2014_train", "coco_2014_val", "coco_2015_test", "visual_genome"],
+    )
+    parser.add_argument(
+        "--dataset-path",
+        help="path to image folder dataset, if not provided expects a detection dataset",
+        default="",
+    )
+    parser.add_argument(
+        "opts",
+        help="Modify config options using the command-line",
+        default=None,
+        nargs=argparse.REMAINDER,
+    )
+    return parser
+
+
+def extract_feature_on_dataset(model, data_loader, dump_folder):
+    for idx, inputs in enumerate(tqdm.tqdm(data_loader)):
+        with torch.no_grad():
+            image_id = inputs[0]["image_id"]
+
+            # compute features and proposals
+            images = model.preprocess_image(inputs)
+            features = model.backbone(images.tensor)
+            proposals, _ = model.proposal_generator(images, features)
+
+            # pooled features and box predictions
+            box_features, pooled_features = model.roi_heads.get_rois_features(
+                features, proposals
+            )
+            predictions = model.roi_heads.box_predictor(pooled_features)
+            cls_probs = F.softmax(predictions[0], dim=-1)
+            cls_probs = cls_probs[:, :-1]  # background is last
+            predictions, r_indices = model.roi_heads.box_predictor.inference(
+                predictions, proposals
+            )
+
+            # postprocess
+            height = inputs[0].get("height")
+            width = inputs[0].get("width")
+            r = postprocessing.detector_postprocess(predictions[0], height, width)
```

Comment:
I haven't seen any major difference in the performance yet compared to old features. However I will check if we can get the features before the box head regression and if it helps the performance.

Reply:
cc @apsdehal Do you know what was the difference in extracting features before and after box regression?

Reply:
I just did a bit digging and think the original one uses the proposal boxes directly before second regression. https://github.com/facebookresearch/mmf/blob/4517e13f8407f95840340e90a6d30df344d08ea9/tools/scripts/features/extract_features_vmb.py#L184

I think this is more than observed performance difference -- if there is any subtle difference we are aware of that can cause a difference, it should be stated at least. And past experience suggest that having the additional box regression does not help, and can sometimes hurt the accuracy. So for a faithful implementation (ideally it should be able to extract the original BUTD features), I still prefer for it to be implemented just as what is being done before.

Reply:
I have added ability to extract features from both last FC (fc7) and second last FC (fc6) layers for FPN and DC5 models. Also now we use proposal bbox instead of predicted boxes exactly similar to old pythia implementation. 

I am still keeping ability to have last layer features because current BERT based models don't finetune last layer and hence should use last layer FC features.